### PR TITLE
Asynchronous implementation of `MemorySource<TSource>.CopyToAsync`

### LINKED ISSUE #265: Asynchronous implementation of `MemorySource<TSource>.CopyToAsync`
### Overview

Currently, the stream returned by `AsStream()` implements `CopyToAsync` which calls the provided `CopyTo` implementation, which calls `destination.Write`. This causes unexpected synchronous I/O.

Due to the synchronous I/O, an exception is thrown if this method is used in ASP.NET Core to write to the response stream:

```
System.InvalidOperationException: Synchronous operations are disallowed. Call WriteAsync or set AllowSynchronousIO to true instead.
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpResponseStream.Write(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.Stream.Write(ReadOnlySpan`1 buffer)
   at Microsoft.Toolkit.HighPerformance.Streams.MemoryStream`1.CopyTo(Stream destination, Int32 bufferSize)
   at Microsoft.Toolkit.HighPerformance.Streams.MemoryStream`1.CopyToAsync(Stream destination, Int32 bufferSize, CancellationToken cancellationToken)
```

### API breakdown

**Note: These are internal APIs**

In order to implement this in a performant way, I'm thinking of adding a property `Memory` to `ISpanOwner`:
```csharp
internal interface ISpanOwner
{
    /// <summary>
    /// Gets a <see cref="Memory{T}"/> instance wrapping the underlying memory area.
    /// </summary>
    Memory<byte> Memory { get; }
}
```

And then implementing `CopyToAsync` like this:
```csharp
partial class MemoryStream<TSource>
{
    /// <inheritdoc/>
    public sealed override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
    {
        if (cancellationToken.IsCancellationRequested)
        {
            return Task.FromCanceled(cancellationToken);
        }

        try
        {
            MemoryStream.ValidateDisposed(this.disposed);

            Memory<byte> source = this.source.Memory.Slice(this.position);

            this.position += source.Length;

            return destination.WriteAsync(source, cancellationToken).AsTask();
        }
        catch (OperationCanceledException e)
        {
            return Task.FromCanceled(e.CancellationToken);
        }
        catch (Exception e)
        {
            return Task.FromException(e);
        }
    }
}
```

### Usage example

```csharp
public async Task ProcessRequestAsync(HttpContext context)
{
    var memory = await _service.GetDataAsync();
    var stream = memory.AsStream();
    await stream.CopyToAsync(context.Response.Body);
}
```

### Breaking change?

No

### Alternatives

Not applicable

### Additional context

Wasn't sure at first if this counted as a bug, upon closer inspection it looks a bit more likely, so I may have chosen the wrong template.

### Help us help you

Yes, I'd like to be assigned to work on this item

### PR DESCRIPTION
**Pull request by OronDF343**
**Closes #265**

This PR replaces the implementation of `MemorySource<TSource>.CopyToAsync` with one that calls `destination.WriteAsync` instead of `destination.Write`. See linked issue for motivation.
To support this change, `ISpanOwner.Memory` has been added and implemented. Feedback appreciated.

## PR Checklist

- [x] Created a feature/dev branch in your fork (vs. submitting directly from a commit on main)
- [x] Based off latest main branch of toolkit
- [x] PR doesn't include merge commits (always rebase on top of our main, if needed)
- [x] Tested code with current [supported SDKs](../#supported)
- [x] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [x] Contains **NO** breaking changes
- [x] Every new API (including internal ones) has full XML docs
- [x] Code follows all style conventions

## Other information

Please advise if tests are applicable for these changes.

### CODE REVIEW

File: CommunityToolkit.HighPerformance/Streams/MemoryStream{TSource}.cs [hunk 1]
```diff
@@ -119,9 +119,13 @@ public sealed override Task CopyToAsync(Stream destination, int bufferSize, Canc
...
         {
-            CopyTo(destination, bufferSize);
+            MemoryStream.ValidateDisposed(this.disposed);
 
-            return Task.CompletedTask;
+            Memory<byte> source = this.source.Memory.Slice(this.position);
+
+            this.position += source.Length;
+
+            return destination.WriteAsync(source, cancellationToken).AsTask();
```
Comment (gfoidl) [inline CommunityToolkit.HighPerformance/Streams/MemoryStream{TSource}.cs:128]: Should the order be swapped?
If any exception occurs, then position won't be updated, which is what's expected?

Comment (OronDF343) [reply CommunityToolkit.HighPerformance/Streams/MemoryStream{TSource}.cs:128]: I implemented this based on the existing `CopyTo` implementation, which itself seems to be inconsistent with the `Read` and `Write` overloads. Should I fix the original within this PR as well?

Comment (Sergio0694) [reply CommunityToolkit.HighPerformance/Streams/MemoryStream{TSource}.cs:128]: The order there is following what `MemoryStream` does in the BCL ([see here](https://github.com/dotnet/runtime/blob/7adbc10dfb865fd5ecc37e309caa2d955b6061b0/src/libraries/System.Private.CoreLib/src/System/IO/MemoryStream.cs#L368)). The `Read` and `Write` APIs read/write first and then update the position, whereas `CopyTo` and `CopyToAsync` first update the position and then defer to the destination stream. We're just following suit with that same logic.

File: CommunityToolkit.HighPerformance/Streams/Sources/ArrayOwner.cs [hunk 2]
```diff
@@ -74,4 +74,14 @@ public Span<byte> Span
...
     }
+
+    /// <inheritdoc/>
+    public Memory<byte> Memory
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        get
+        {
+            return this.array.AsMemory(this.offset, this.length);
+        }
```
Comment (Sergio0694) [inline CommunityToolkit.HighPerformance/Streams/Sources/ArrayOwner.cs:85]: Nit: use a one liner.
```suggestion
        get => this.array.AsMemory(this.offset, this.length);
```

File: CommunityToolkit.HighPerformance/Streams/Sources/MemoryManagerOwner.cs [hunk 3]
```diff
@@ -61,4 +61,14 @@ public Span<byte> Span
...
     }
+
+    /// <inheritdoc/>
+    public Memory<byte> Memory
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        get
+        {
+            return this.memoryManager.Memory.Slice(this.offset, this.length);
+        }
```
Comment (Sergio0694) [inline CommunityToolkit.HighPerformance/Streams/Sources/MemoryManagerOwner.cs:72]: Use a one liner here as well.
```suggestion
        get => this.memoryManager.Memory.Slice(this.offset, this.length);
```

**Review by Sergio0694 — CHANGES_REQUESTED** [review]
Could you add some unit tests too? 🙂

**Comment by OronDF343** [comment]
I've added basic unit tests

**Pull request merged**